temporian/core/data/dtype.py

```diff
@@ -38,3 +38,45 @@ def MissingValue(dtype: DType) -> Any:
         return ""
     else:
         raise ValueError(f"Non implemented type {dtype}")
+
+
+def get_resulting_dtype(dtype1: DType, dtype2: DType) -> DType:
+    """
+    Get the resulting dtype when combining two arrays with the specified dtypes.
+
+    This function takes two dtypes as input and returns the expected
+    output dtype based on a simplified dtype hierarchy.
+
+    Args:
+        dtype1: The dtype of the first array.
+        dtype2: The dtype of the second array.
+
+    Returns:
+        DType: The resulting dtype
+
+    Raises:
+        ValueError: If either of the input dtypes is not supported.
+    """
+    dtype_hierarchy = {
+        INT32: 1,
+        INT64: 2,
+        FLOAT32: 3,
+        FLOAT64: 4,
+    }
+
+    if dtype1 not in dtype_hierarchy or dtype2 not in dtype_hierarchy:
+        supported_dtypes = ", ".join(dtype_hierarchy.keys())
+        raise ValueError(
+            f"Invalid dtype(s): [{dtype1}, {dtype2}]. Supported dtypes:"
+            f" {supported_dtypes}"
+        )
+
+    # Find the highest hierarchy value between the two input dtypes
+    max_hierarchy = max(dtype_hierarchy[dtype1], dtype_hierarchy[dtype2])
```

Comment:
int32 => float32 or int64 => float32 is possibly a drop in precision.

I like this idea, but what about adding it to our next sync for the details?
For example, why not int32=>float32 and int64=>float64, or doing something like numpy (your link: https://numpy.org/doc/stable/reference/generated/numpy.result_type.html), or possibly just disallowing operations between different types (like in go.